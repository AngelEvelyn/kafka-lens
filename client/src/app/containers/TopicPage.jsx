import PropTypes from 'prop-types';
import React from 'react';
import { ipcRenderer } from 'electron';
import Topic from '../components/Topic';
import PartitionInfo from '../components/PartitionInfo';
import RouteBar from '../components/RouteBar';
import MessageInfo from '../components/MessageInfo';
import MessageList from '../components/MessageList';
import LoadingData from '../components/LoadingData';

import '../css/TopicPage.scss';
import '../css/PartitionList.scss';
import lensIcon from '../../../../assets/images/lens-icon.png';
import logger from '../../utils/logger';

class TopicPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentTopicMetadata: {},
      currentPartitionMetadata: {},
      topicName: '',
      messages: [],
      partitionId: '',
      lastElement: '',
      showingPartitionMetadata: false,
      loadingData: false,
    };

    this.showPartitions = this.showPartitions.bind(this);
    this.showMessages = this.showMessages.bind(this);
  }
  // Lifecycle methods

  componentDidMount() {
    ipcRenderer.on('partition:getMessages', (e, messages) => {
      this.setState({ messages });
    });

    // This will get an object from the main process with the partition data
    // includes: highwaterOffset, earliestOffset, messageCount, leader, and replicas
    ipcRenderer.on('partition:getData', (e, data) => {
      this.setState({ currentPartitionMetadata: data });
    });
  }

  // Called when topic is clicked in order to show partitions
  showPartitions(event) {
    const { topicList } = this.props;
    const topicName = event.target.getAttribute('topicname');
    const topicIndex = parseInt(event.target.id, 10);
    const topicInfo = topicList[topicIndex];

    topicInfo.showPartitions = !topicInfo.showPartitions;

    const { uri } = this.props;
    ipcRenderer.send('partition:getMessages', {
      kafkaHostURI: uri,
      topicName,
    });

    const newState = this.state;

    const { showingPartitionMetadata } = this.state;
    if (showingPartitionMetadata === true) newState.showingPartitionMetadata = false;

    newState.currentTopicMetadata = topicInfo;
    newState.loadingData = true;

    return this.setState(newState);
  }

  showMessages(event) {
    const newTopicName = event.target.getAttribute('topicname');
    logger.log('newTopicName from the partition div:', newTopicName);
    const newPartitionId = parseInt(event.target.id, 10);
    logger.log('newPartitionId:', newPartitionId);

    const element = event.target;
    const { lastElement, partitionId, topicName } = this.state;

    const { uri } = this.props;
    const kafkaHostURI = uri;

    if (lastElement !== element) {
      if (lastElement !== '') {
        lastElement.classList.remove('highlight-this');
      }

      ipcRenderer.send('partition:getMessages', {
        kafkaHostURI,
        topicName: newTopicName,
        partitionId,
      });

      ipcRenderer.send('partition:getData', {
        kafkaHostURI,
        partitionId,
        topicName: newTopicName,
      });
      this.setState({
        lastElement: element,
      });

      element.classList.add('highlight-this');
    }

    if (newPartitionId !== partitionId || newTopicName !== topicName) {
      this.setState({
        messages: [],
        topicName: newTopicName,
        partitionId: newPartitionId,
        showingPartitionMetadata: true,
      });
    }
  }

  render() {
    const { topicList, isConnected, uri } = this.props;
    const {
      showingPartitionMetadata,
      showPartitions,
      currentPartitionMetadata,
      partitionId,
      currentTopicMetadata,
      topicName,
      messages,
      loadingData,
    } = this.state;

    const shouldDisplayPartitionMetadata =
      showingPartitionMetadata === true &&
      Object.keys(currentPartitionMetadata).length &&
      messages.length > 0;

    const shouldDisplayMessageInfo = showingPartitionMetadata === true && messages.length > 0;

    const Topics = topicList.map((topicInfo, i) => (
      <Topic
        key={topicInfo.topicName}
        id={i}
        topicInfo={topicInfo}
        showPartitions={this.showPartitions}
        shouldDisplayPartitions={showPartitions}
        showMessages={this.showMessages}
      />
    ));

    const connected = <h5 className="connection-header">Connected</h5>;
    const disconnected = <h5 className="disconnected-header">Disconnected</h5>;

    return (
      <div className="grid-container">
        <div className="title-bar">Kafka Lens</div>
        <div className="route-bar">
          {logger.log('topicInfo:', currentTopicMetadata)}
          <RouteBar
            partitionId={partitionId}
            topicName={currentTopicMetadata.topicName}
            showingPartitionMetadata={showingPartitionMetadata}
          />
        </div>
        <div className="logo-box">
          <img className="lens-icon" src={lensIcon} alt="kakfa lens logo" />
        </div>
        <div className="topics-header">Topics</div>
        <div className="partition-info">
          {loadingData && messages.length === 0 && <LoadingData />}
          {shouldDisplayPartitionMetadata && (
            <PartitionInfo partitionInfo={currentPartitionMetadata} partitionId={partitionId} />
          )}
        </div>
        {logger.log('messages:', messages)}
        <div className="message-info">
          {shouldDisplayMessageInfo && <MessageInfo lastMessage={messages[0]} />}
        </div>
        <div className="list-display">{Topics}</div>
        <div className="message-box">
          <MessageList topicName={topicName} messageArray={messages} />
        </div>
        <div className="connection-status">
          {isConnected === true ? connected : disconnected}
          <div className="connection-uri">{uri}</div>
        </div>
      </div>
    );
  }
}

export default TopicPage;

TopicPage.propTypes = {
  isConnected: PropTypes.bool.isRequired,
  topicList: PropTypes.arrayOf(PropTypes.object).isRequired,
  uri: PropTypes.string.isRequired,
};
