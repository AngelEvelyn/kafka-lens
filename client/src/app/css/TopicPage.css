.partition-list {
  margin-top: 25px;
}

.topic-list {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: 1fr;
}

.bottom-container {
  display: grid;
  grid-template-columns: 1fr 1fr;
}
