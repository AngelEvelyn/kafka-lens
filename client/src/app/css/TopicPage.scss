@import 'theme';


.grid-container {
  display: grid;
  grid-template-columns: 0.8fr 0.9fr 1.3fr 1fr;
  grid-template-rows: 0.3fr 0.4fr 1.8fr 1.5fr 1fr 1fr 1fr;
  grid-template-areas: "title-bar title-bar title-bar title-bar" "navi-bar route-bar route-bar route-bar" "navi-bar more-info-box message-box message-box" "navi-bar more-info-box message-box message-box" "navi-bar more-info-box message-box message-box" "navi-bar health-box metrics-box metrics-box" "navi-bar health-box metrics-box metrics-box";
  min-width: 100vw;
  min-height: 100vh;
  max-width: 100vw;
  max-height: 100vh;
}

.title-bar { 
  grid-area: title-bar; 
  background-color: $new-bg-blue;
  color: $base-font-color;
  text-align: center;
  padding-top: 5px;
}

.navi-bar {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr 0.3fr 2.1fr 0.6fr;
  grid-template-areas: "logo-box" "topics-header" "list-display" "connection-status";
  grid-area: navi-bar;
  background-color: $new-bg-blue;
  color: $base-font-color;
}

.logo-box { 
  grid-area: logo-box; 
  text-align: center;
}

.lens-icon {
  height: 100px;
  width: 100px;
}

.topics-header { 
  grid-area: topics-header; 
  font-size: 32px;
  padding-left: 29px;
}

.list-display { 
  grid-area: list-display; 
  overflow: auto;
  padding-left: 44px;
  padding-top: 18px;
}

.connection-status { 
  grid-area: connection-status; 
  text-align: center;
  padding-top: 15px;
}

.connection-header {
  font-size: 24px;
  color: #76C652;
}

.connection-uri {
  color: $base-font-color;
}

.route-bar { 
  grid-area: route-bar; 
  background-color: $highlight-bar-color;
}

.more-info-box { 
  grid-area: more-info-box; 
  background-color: $second-bg-color; 
}

.message-box { 
  grid-area: message-box; 
}

.health-box { 
  grid-area: health-box; 
  background-color: $highlight-bar-color;
}

.metrics-box { 
  grid-area: metrics-box; 
  background-color: $metrics-background-color;
  text-align: center;
}

.metric-demo {
  height: 100%;
  width: 75%;
}

::-webkit-scrollbar {     
  background-color: $new-bg-blue;
  width: .8em
}

::-webkit-scrollbar-thumb:window-inactive,
::-webkit-scrollbar-thumb {
  background:  $base-font-color;
}








// .grid-container {
//   display: grid;
//   grid-template-columns: 0.8fr 1fr 1.2fr 1fr;
//   grid-template-rows: 0.3fr 0.3fr 1.3fr 1.3fr 1.8fr;
//   grid-template-areas: "title-bar title-bar title-bar title-bar" "navi-bar route-bar route-bar route-bar" "navi-bar info-box message-display message-display" "navi-bar info-box message-display message-display" "navi-bar health-box metrics metrics";
//   min-width: 100vw;
//   min-height: 100vh;
//   max-width: 100vw;
//   max-height: 100vh;
// }

// .title-bar { 
//   grid-area: title-bar; 
//   background-color: $new-bg-blue;
//   color: $base-font-color;
//   text-align: center;
//   padding-top: 5px;
// }

// .navi-bar {
//   display: grid;
//   grid-template-columns: 1fr 1fr 1fr;
//   grid-template-rows: 1.6fr 0.4fr;
//   grid-template-areas: "topic-list topic-list topic-list" "connection-status connection-status connection-status";
//   grid-area: navi-bar;
//   background-color: $new-bg-blue;
// }

// .topic-list { 
//   grid-area: topic-list; 
// }

// .connection-status { 
//   grid-area: connection-status; 
// }

// .route-bar { 
//   grid-area: route-bar; 
//   background-color: $highlight-bar-color;
// }

// .info-box { 
//   grid-area: info-box;
//   background-color: $second-bg-color; 
// }

// .message-display { 
//   grid-area: message-display; 
// }

// .metrics { 
//   grid-area: metrics; 
//   background-color: $metrics-background-color;
// }

// .health-box { 
//   grid-area: health-box; 
//   background-color: $highlight-bar-color;
// }








// .wrapper {
//   display: grid;
//   grid-template-columns: 1fr 1fr 3fr;
//   min-width: 100vw;
//   min-height: 100vh;
//   max-width: 100vw;
//   max-height: 100vh;
//   // height: 100%;
// }

// .side-navigation {
//   display: grid;
//   grid-row: 1/4;
//   grid-template-rows: 1fr 3fr;
//   background-color: $new-bg-blue;
//   color: $base-font-color;
//   // position: absolute;
// }

// .kafka-logo-box {
//   text-align: center;
//   padding-top: 36px;
// }

// .lens-icon {
//   height: 100px;
//   width: 100px;
// }

// .topic-partition-list {
//   position: absolute;
//   top: 188px;
//   height: 100%;
//   overflow: auto;
//   padding-left: 3%;
//   padding-right: 2%;
//   // overflow: auto;
//   // padding-left: 20px;
//   // padding-right: 20px;
// }

// ::-webkit-scrollbar {     
//   background-color: $new-bg-blue;
//   width: .8em
// }

// ::-webkit-scrollbar-thumb:window-inactive,
// ::-webkit-scrollbar-thumb {
//   background:  $base-font-color;
// }

// .more-info-box {
//   grid-row: 1/4;
// }

// .messages-box {
//   grid-column: 3/4;
//   grid-row: 1/3;
// }

// .metrics-box {
//   grid-column: 3/4;
//   grid-row: 3/4;
// }




// @import 'theme';

// h3 {
//   text-align: center;
// }

// .topic-list {
//   display: grid;
//   grid-template-columns: 1fr 1fr 1fr;
//   grid-template-rows: 1fr;
//   grid-row-gap: 5em;
//   text-align: center;
//   padding-top: 67px;
//   padding-bottom: 81px;
//   border-bottom: 1px solid lightgray;
// }

// .bottom-container {
//   display: grid;
//   grid-template-columns: 1fr 1fr;
//   padding-top: 30px;
// }

// .message-list {
//   justify-self: center;
// }

// .highlight-topic {
//   background-color: gray;
// }

// .highlight-this {
//   background-color: #f7ae4f;
//   color: white;
//   border: $border-highlight;
// }

// incoming messages loading animation

// .spinner {
//   margin: 50px auto 0;
//   padding-bottom: 20px;
//   width: 70px;
//   text-align: center;
// }

// .spinner > div {
//   width: 14px;
//   height: 14px;
//   background-color: $dark-font-primary;

//   border-radius: 100%;
//   display: inline-block;
//   -webkit-animation: sk-bouncedelay 1.4s infinite ease-in-out both;
//   animation: sk-bouncedelay 1.4s infinite ease-in-out both;
// }

// .spinner .bounce1 {
//   -webkit-animation-delay: -0.32s;
//   animation-delay: -0.32s;
// }

// .spinner .bounce2 {
//   -webkit-animation-delay: -0.16s;
//   animation-delay: -0.16s;
// }

// @-webkit-keyframes sk-bouncedelay {
//   0%, 80%, 100% { -webkit-transform: scale(0) }
//   40% { -webkit-transform: scale(1.0) }
// }

// @keyframes sk-bouncedelay {
//   0%, 80%, 100% { 
//     -webkit-transform: scale(0);
//     transform: scale(0);
//   } 40% { 
//     -webkit-transform: scale(1.0);
//     transform: scale(1.0);
//   }
// }


