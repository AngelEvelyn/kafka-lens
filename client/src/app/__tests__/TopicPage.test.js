import React from 'react';
import { shallow } from 'enzyme';

import TopicPage from '../containers/TopicPage';

describe('TopicPage.js unit tests', () => {
  let wrapper;
  const props = {
    topicList: []
  };
  beforeAll(() => {
    wrapper = shallow(<TopicPage {...props} />);
  });

  it('Should render without errors', () => {
    expect(wrapper).toMatchSnapshot();
    console.log('wrapppererereprerperper', wrapper);
  });

  it('Should render Topic components if there are topics in state', () => {
    expect(wrapper.exists('Topic')).toBe(false);
    wrapper.setProps({
      topicList: [{ message: 'test1' }, { message: 'test2' }, { message: 'test3' }]
    });
    expect(wrapper.exists('Topic')).toBe(true);
  });

  xit('Should render loading animation if this.state.messages.length > 0', () => {
    const instance = wrapper.instance();
    expect(
      wrapper.containsMatchingElement(
        <div class="spinner">
          <div class="bounce1" />
          <div class="bounce2" />
          <div class="bounce3" />
        </div>
      )
    ).toBeFalsy();
    instance.setState({ messages: [1, 2, 3, 4, 5] });
    expect(
      wrapper.containsMatchingElement(
        <div class="spinner">
          <div class="bounce1" />
          <div class="bounce2" />
          <div class="bounce3" />
        </div>
      )
    ).toBeTruthy();
  });
});
