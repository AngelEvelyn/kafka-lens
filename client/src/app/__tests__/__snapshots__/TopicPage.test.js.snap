// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicPage.js unit tests Should render without errors 1`] = `
<div
  className="grid-container"
>
  <div
    className="title-bar"
  >
    Kafka Lens
  </div>
  <div
    className="navi-bar"
  >
    <div
      className="logo-box"
    >
      <img
        className="lens-icon"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <div
      className="topics-header"
    >
      Topics
    </div>
    <div
      className="list-display"
    />
    <div
      className="connection-status"
    >
      <div
        className="connection-header"
      >
        <h5
          className="disconnected-header"
        >
          Disconnected
        </h5>
      </div>
      <div
        className="connection-uri"
      />
    </div>
  </div>
  <div
    className="route-bar"
  >
    <RouteBar
      partitionNumber={-1}
    />
  </div>
  <div
    className="more-info-box"
  />
  <div
    className="message-box"
  >
    <MessageList
      messageArray={Array []}
    />
  </div>
  <div
    className="health-box"
  />
  <div
    className="metrics-box"
  >
    <img
      className="metric-demo"
      src={
        Object {
          "process": [Function],
        }
      }
    />
  </div>
</div>
`;
